src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs

```diff
@@ -380,6 +357,53 @@ private static bool SymbolsAreCompatibleCore(ISymbol symbol, ISymbol newSymbol,
             return false;
         }
 
+        private static bool CompareAcrossSemanticModels(ISymbol symbol, ISymbol newSymbol)
+        {
+            // SymbolEquivalenceComparer performs Location equality checks for locals, labels, range-variables and local
+            // functions. As we are comparing symbols from different semantic models, locations will differ. Hence
+            // perform minimal checks for these symbol kinds.
+
+            if (symbol.Kind != newSymbol.Kind)
+                return false;
+
+            if (symbol is ILocalSymbol localSymbol && newSymbol is ILocalSymbol newLocalSymbol)
+            {
+                return newSymbol.IsImplicitlyDeclared == symbol.IsImplicitlyDeclared &&
+                       symbol.Name == newSymbol.Name &&
+                       CompareAcrossSemanticModels(localSymbol.Type, newLocalSymbol.Type);
+            }
+
+            if (symbol is ILabelSymbol && newSymbol is ILabelSymbol)
+                return symbol.Name == newSymbol.Name;
+
+            if (symbol is IRangeVariableSymbol && newSymbol is IRangeVariableSymbol)
+                return symbol.Name == newSymbol.Name;
+
+            if (symbol is IParameterSymbol parameterSymbol &&
+                newSymbol is IParameterSymbol newParameterSymbol &&
+                parameterSymbol.ContainingSymbol.IsAnonymousOrLocalFunction() &&
+                newParameterSymbol.ContainingSymbol.IsAnonymousOrLocalFunction())
+            {
+                return symbol.Name == newSymbol.Name &&
+                       parameterSymbol.IsRefOrOut() == newParameterSymbol.IsRefOrOut() &&
+                       CompareAcrossSemanticModels(parameterSymbol.Type, newParameterSymbol.Type);
```

Comment:
❔ Did this change (from `SymbolEquivalenceComparer.Instance.Equals(param.Type, newParam.Type)`) fix a bug? If so, I'm curious which one. Feel free to put the answer in reply to this comment and not in code, since a test will already be covering regressions.